### PR TITLE
Table Updates

### PR DESCRIPTION
**Pull request by shaneeza**
## This PR includes:

- updated`useLeafyGreenTable` hook
- adds `useLeafyGreenVirtualTable` hook
- remove mapping of children from inside the [InternalRowWithRT](packages/table/src/Row/InternalRowWithRT.tsx) component
- memorizes rows inside [InternalRowWithRT](packages/table/src/Row/InternalRowWithRT.tsx)
- supports styled components
- multiline row support
- removal of table v10 and it's adapter

## Not include:

- ForwardRefs. #2539 
- Documentation and examples - PR coming soon
- Updated specs and test utils - PR coming soon
- sticky header scroll shadow - PR coming soon
- changesets

## Storybook links
_Note_: Some stories with hundreds of rows will freeze storybook on initial render. A workaround while I figure this out is to view the story outside of an iframe. 
_Also_, tables currently expand faster and scroll better in incognito

🎟 _Jira ticket:_ [LG-4414](https://jira.mongodb.org/browse/LG-4414)

### For bug fixes, new features & breaking changes

- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] I have run `yarn changeset` and documented my changes

### For new components

- [ ] I have added my new package to the global tsconfig
- [ ] I have added my new package to the Table of Contents on the global README
- [ ] I have verified the Live Example looks as intended on the design website.

### CODE REVIEW

File: packages/table/src/Table/Table.tsx [hunk 1]
```diff
@@ -26,14 +33,28 @@ const Table = forwardRef<HTMLDivElement, TableProps<any>>(
...
       table,
-      disableAnimations = false,
       'data-lgid': lgidProp = LGIDS.root,
       ...rest
     }: TableProps<T>,
     containerRef: ForwardedRef<HTMLDivElement>,
   ) => {
     const baseFontSize: BaseFontSize = useUpdatedBaseFontSize(baseFontSizeProp);
     const { theme, darkMode } = useDarkMode(darkModeProp);
+    //TODO: find a better way to do all these checks
```
Comment (shaneeza) [inline packages/table/src/Table/Table.tsx:43]: maybe make a hook that returns all of this

File: packages/table/src/useLeafyGreenTable/useLeafyGreenTable.tsx [hunk 2]
```diff
@@ -69,25 +77,39 @@ function useLeafyGreenTable<T extends LGRowData, V extends unknown = unknown>({
...
+  for (let i = 0; i < rowsCopy.length; i++) {
+    if (
+      rowsCopy[i].original.renderExpandedContent &&
+      rowsCopy[i].getIsExpanded()
+    ) {
+      rowsCopy.splice(i + 1, 0, {
+        ...rowsCopy[i],
+        id: `${rowsCopy[i].id}-expandedContent`,
+        original: {
+          // TODO: move outside of original
```
Comment (shaneeza) [inline packages/table/src/useLeafyGreenTable/useLeafyGreenTable.tsx:100]: TODO

File: packages/table/package.json [hunk 3]
```diff
@@ -33,10 +33,14 @@
     "@leafygreen-ui/tokens": "^2.11.0",
     "@leafygreen-ui/typography": "^19.2.0",
     "@lg-tools/test-harnesses": "^0.1.2",
-    "@tanstack/react-table": "^8.13.2",
+    "@tanstack/react-table": "^8.20.5",
+    "@tanstack/react-virtual": "^3.10.7",
     "lodash": "^4.17.21",
     "polished": "^4.2.2",
-    "react-virtual": "^2.10.4"
+    "react-virtual": "^2.10.4",
```
Comment (shaneeza) [inline packages/table/package.json:40]: will be removed in another PR

File: packages/table/package.json [hunk 4]
```diff
@@ -33,10 +33,14 @@
...
+    "react-fast-compare": "3.2.2",
+    "react-transition-group": "^4.4.5",
+    "react-keyed-flatten-children": "^1.3.0"
```
Comment (shaneeza) [inline packages/table/package.json:43]: 42 and 43 will be removed in another PR

File: packages/table/src/Cell/Cell.stories.tsx
Comment (shaneeza) [inline packages/table/src/Cell/Cell.stories.tsx:1]: add a depth story, needs row context

File: packages/table/src/Row/Row.tsx [hunk 5]
```diff
@@ -14,17 +17,39 @@ import { RowContextProvider } from './RowContext';
...
+  const { shouldAlternateRowColor } = useTableContext();
+  const { virtualTable } = useVirtualTableContext();
+
   return (
-    <RowContextProvider disabled={disabled}>
+    <>
       {row ? (
-        <InternalRowWithRT row={row} virtualRow={virtualRow} {...rest} />
+        <MemoizedInternalRowWithRT
+          // @ts-ignore FIXME:
```
Comment (shaneeza) [inline packages/table/src/Row/Row.tsx:31]: This are fixed in this PR #2539

File: packages/table/src/Row/InternalRowWithRT.tsx [hunk 6]
```diff
@@ -28,67 +18,79 @@ const InternalRowWithRT = <T extends LGRowData>({
...
       </InternalRowBase>
-    </Polymorph>
+    </RowContextProvider>
   );
 };
 
 export default InternalRowWithRT;
+
+// @ts-expect-error FIXME: the types are generic
+const arePropsEqual = (prevProps, nextProps) => {
```
Comment (shaneeza) [inline packages/table/src/Row/InternalRowWithRT.tsx:83]: Addressed in this PR #2539

File: packages/table/src/Cell/Cell.tsx [hunk 7]
```diff
@@ -1,44 +1,28 @@
...
-          { [disableAnimationStyles]: disableAnimations },
-          contentClassName,
-        )}
-      >
-        {children}
-      </div>
-    </td>
+    <>
+      {!cell && (
+        <InternalCellWithoutRT {...rest}>{children}</InternalCellWithoutRT>
```
Comment (TheSonOfThomp) [inline packages/table/src/Cell/Cell.tsx:17]: This is confusing lol ("If cell does not exist, render a cell")

File: packages/table/src/Cell/Cell.tsx [hunk 8]
```diff
@@ -1,44 +1,28 @@
...
+      )}
+
+      {cell && (
```
Comment (TheSonOfThomp) [inline packages/table/src/Cell/Cell.tsx:20]: Could use ternary here
```tsx
cell ? <InternalCellWithRT /> : <InternalCellWithoutRT />
```

File: packages/table/src/Cell/HeaderCell/HeaderCell.tsx [hunk 9]
```diff
@@ -33,16 +27,14 @@ const HeaderCell = <T extends LGRowData>({
...
-  const isSelectable = !!table && !!table.hasSelectableRows;
+  const { isSelectable } = useTableContext();
 
   let columnName, sortState, onSortIconClick;
 
-  if (header && header.column.getCanSort()) {
+  if (header && header?.column.columnDef.enableSorting) {
     columnName = header.column.columnDef.header as string;
     const headerSortDirection = header.column.getIsSorted().toString();
     sortState = HeaderSortState[headerSortDirection];
```
Comment (TheSonOfThomp) [inline packages/table/src/Cell/HeaderCell/HeaderCell.tsx:40]: I'd create a utility for this 
```suggestion
  const {columnName, sortState, onSortIconClick} = getHeaderCellState(header)
```

File: packages/table/src/ExpandedContent/ExpandedContent.tsx [hunk 10]
```diff
@@ -1,61 +1,46 @@
...
-
   return (
-    <InternalRowBase {...rest} aria-hidden={!isExpanded}>
+    <InternalRowBase
+      {...rest}
+      ref={node => {
+        // TODO: fix me
+        // This gets the dynamic size of the element
+        if (virtualTable) virtualTable.measureElement(node);
+      }}
```
[the remark this answers is not in the document]
Comment (shaneeza) [reply packages/table/src/ExpandedContent/ExpandedContent.tsx:34]: Updated in this PR https://github.com/mongodb/leafygreen-ui/pull/2539/files#diff-0dfe10c35fa37d766d24e5444f598c29c9ed027ed2821cc557598175064f3205R75

File: packages/table/src/Row/Row.styles.ts [hunk 11]
```diff
@@ -88,3 +88,21 @@ export const disabledStyles: Record<Theme, string> = {
...
+  isVirtualRow: boolean,
+  isDisabled: boolean,
+  isExpanded: boolean,
+  theme: Theme,
+) =>
+  cx({
+    [grayZebraRowStyles[theme]]:
+      isOddVSRow && shouldAlternateRowColor && !isSelected,
+    [zebraStyles[theme]]:
+      !isVirtualRow && shouldAlternateRowColor && !isSelected,
```
[the remark this answers is not in the document]
Comment (shaneeza) [reply packages/table/src/Row/Row.styles.ts:105]: I only copied these from another file, but it looks like `grayZebraRowStyles` is only for a virtual table while `zebraStyles` is for a regular table.

File: packages/table/src/Row/Row.tsx [hunk 12]
```diff
@@ -14,17 +17,39 @@ import { RowContextProvider } from './RowContext';
...
+  const { theme } = useDarkMode();
+  const { shouldAlternateRowColor } = useTableContext();
+  const { virtualTable } = useVirtualTableContext();
+
   return (
-    <RowContextProvider disabled={disabled}>
+    <>
       {row ? (
-        <InternalRowWithRT row={row} virtualRow={virtualRow} {...rest} />
+        <MemoizedInternalRowWithRT
```
[the remark this answers is not in the document]
Comment (shaneeza) [reply packages/table/src/Row/Row.tsx:30]: Yes, without this, every row rerenders when a row is expanded. Why do you think it does not need to be memoized?

File: packages/table/src/Row/Row.types.ts [hunk 13]
```diff
@@ -20,9 +20,40 @@ export interface InternalRowWithRTProps<T extends LGRowData>
...
+  /**
+   * Determines whether alternating rows will have dark backgrounds.
+   * @default false
+   */
+  shouldAlternateRowColor: boolean;
+
+  /**
+   * Whether the theme is dark or light
+   */
+  theme: Theme;
```
[the remark this answers is not in the document]
Comment (shaneeza) [reply packages/table/src/Row/Row.types.ts:36]: This is related to what i mentioned above, memoizing does not work if i call context within the component so instead of calling `const { theme } = useDarkMode();` inside of `MemoizedInternalRowWithRT`, i pass it as a prop.

File: packages/table/src/TableContext/VirtualTableContext.tsx [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+import { type VirtualTableContextValues } from './TableContext.types';
+
+export const VirtualTableContext = createContext<
+  Partial<VirtualTableContextValues<LGRowData>>
+>({});
+
+export const useVirtualTableContext = <T extends LGRowData>() =>
+  useContext<VirtualTableContextValues<T>>(
+    VirtualTableContext as React.Context<VirtualTableContextValues<T>>,
+  );
```
[the remark this answers is not in the document]
Comment (shaneeza) [reply packages/table/src/TableContext/VirtualTableContext.tsx:19]: They don't have to be; I'll combine them.

File: packages/table/src/useLeafyGreenVirtualTable/useLeafyGreenVirtualTable.tsx [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+  const table = useLeafyGreenTable({
+    data,
+    columns,
+    withPagination,
+    allowSelectAll,
+    hasSelectableRows,
+    ...rest,
+  });
+
+  const { rows } = table;
```
[the remark this answers is not in the document]
Comment (shaneeza) [reply packages/table/src/useLeafyGreenVirtualTable/useLeafyGreenVirtualTable.tsx]:
```
const getRows = useCallback(() => table.getRowModel().rows, [table]
```
and then using like this:
```
getRows()
```
causes the table virtual scrolling to get all wonky. I think it as to do with the useCallback.

I think row is supposed to to be called on every hook call because we need the latest row data to pass to the `useVirtualizer` instance

File: packages/table/CHANGELOG.md [hunk 16]
```diff
@@ -1,5 +1,17 @@
...
+
+### Patch Changes
+
+- Add `react-keyed-flatten-children` to fix install error
+
+## 13.0.0-beta.0
+
+### Major Changes
+
+- First beta pre-release of Table v13. This release is a WIP and should not be used in production.
```
[the remark this answers is not in the document]
Comment (shaneeza) [reply packages/table/CHANGELOG.md:13]: why do we still need the changeset files if we already published?

File: packages/table/src/Table/Table.tsx [hunk 17]
```diff
@@ -33,11 +33,8 @@ const Table = forwardRef<HTMLDivElement, TableProps<any>>(
     const baseFontSize: BaseFontSize = useUpdatedBaseFontSize(baseFontSizeProp);
     const { theme, darkMode } = useDarkMode(darkModeProp);
 
-    const isVirtual =
-      table && (table as LeafyGreenVirtualTable<T>).virtual ? true : false;
-    const virtualTable = isVirtual
-      ? (table as LeafyGreenVirtualTable<T>)!.virtual
-      : undefined;
+    const isVirtual = Boolean((table as LeafyGreenVirtualTable<T>)?.virtual);
+    const virtualTable = isVirtual ? table!.virtual : undefined;
```
Comment (TheSonOfThomp) [inline packages/table/src/Table/Table.tsx:37]: Why `table!`? We should be able to assert that `table` is defined, no?

File: packages/table/src/TableBody/utils/useVirtualScrollPadding.ts [hunk 18]
```diff
@@ -0,0 +1,28 @@
+import { LGRowData } from '../../useLeafyGreenTable';
+import { LeafyGreenVirtualTable } from '../../useLeafyGreenVirtualTable';
+
+export const useVirtualScrollPadding = <T extends LGRowData>(
```
Comment (TheSonOfThomp) [inline packages/table/src/TableBody/utils/useVirtualScrollPadding.ts:4]: This could use some TSDoc

File: packages/table/src/TableContext/TableContext.types.ts [hunk 19]
```diff
@@ -31,20 +16,20 @@ export interface BaseTableContextValue<T extends LGRowData> {
    * Whether rows in the table are selectable
    */
   isSelectable?: boolean;
+
+  /**
+   * Available [properties and methods](https://tanstack.com/virtual/latest/docs/api/virtualizer#virtualizer-instance) returned from the Virtualizer instance.
+   */
+  virtualTable?: LeafyGreenVirtualTable<T>['virtual'];
```
Comment (TheSonOfThomp) [inline packages/table/src/TableContext/TableContext.types.ts:23]: I thought we didn't want to link out to TanStack docs?

**Review by TheSonOfThomp — APPROVED** [review]
Just a few more non-blocking nits

File: packages/table/CHANGELOG.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (TheSonOfThomp) [reply packages/table/CHANGELOG.md:13]: The beta version is already published, but `13.0.0` doesn't exist yet. We need the changeset file to publish that version

File: packages/table/src/Row/Row.types.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (TheSonOfThomp) [reply packages/table/src/Row/Row.types.ts:36]: I'm curious whether memoizing the components is still necessary after the other perf improvements you've made

File: packages/table/src/Row/Row.tsx (hunk 12 above)
[the remark this answers is not in the document]
Comment (TheSonOfThomp) [reply packages/table/src/Row/Row.tsx:29]: Ah, ok. I wasn't sure if after the other perf. improvements you've made, the memoization was no longer necessary

File: packages/table/src/TableContext/TableContext.types.ts (hunk 19 above)
Comment (shaneeza) [reply packages/table/src/TableContext/TableContext.types.ts:23]: i'll sort everything out in my documentation PR
